web-ng/src/app/services/router/router.service.ts

```diff
@@ -137,27 +138,37 @@ export class NavigationService {
    * Navigate to the current URL, replacing the single URL fragment param
    * with the specified value.
    */
-  private setFragmentParam(key: string, value: string | null) {
-    if (value) {
-      this.setFragmentParams(this.getFragmentParams().set(key, value));
-    } else {
-      this.setFragmentParams(this.getFragmentParams().delete(key));
+  private setFragmentParam(updates: Map<string, string | null>) {
+    let params = this.getFragmentParams();
+    for (const [key, value] of updates) {
+      if (value) {
+        params = params.set(key, value);
+      } else {
+        params = params.delete(key);
+      }
     }
+    this.setFragmentParams(params);
+  }
+
+  /**
+   * Get current feature id in the URL fragment.
+   */
+  getFeatureId(): string | null {
+    return this.getFragmentParams().get(
+      NavigationService.FEATURE_ID_FRAGMENT_PARAM
+    );
   }
 
   /**
    * Navigate to the current URL, updating the feature id in the URL fragment.
    */
-  setFeatureId(id: string | null): boolean {
-    if (
-      this.getFragmentParams().get(
-        NavigationService.OBSERVATION_ID_FRAGMENT_PARAM
-      ) !== null
-    ) {
-      return false;
+  setFeatureId(id: string | null) {
+    const updates: { [key: string]: string | null } = {};
+    updates[NavigationService.FEATURE_ID_FRAGMENT_PARAM] = id;
+    if (this.getObservationId()) {
+      updates[NavigationService.OBSERVATION_ID_FRAGMENT_PARAM] = null;
```

Comment:
Rather than explicitly clear the observation id, can we just completely overwrite the fragment params in this case? I don't think there are any params we'd want to carry forward.

Reply:
Sorry I don't quite get the suggestion here. What does `completely overwrite the fragment params` mean?

Reply:
The code here is attempting to set and clear individual values on the URL fragment (`#f=foo&o=bar`). The way it's written, other arbitrary URL fragment params will be preserved by the operation (`#f=foo&o=bar&m=baz` -> `#f=foo2&m=baz`), but we really don't need to do this; selecting a feature presenting clears _all_ fragment params (`#f=foo&o=bar&m=baz` -> `#f=foo2`). So we don't need to override the observation param; we just need to update the fragment params to the new feature id. Does that make sense?

Reply:
Make sense, thanks for explaining. I think we should do it for other fragment params as well. For layer fragment, it can just completely overwrite. The observation fragment is a little bit tricky, changing `#o` should preserve `#f`, can we also completely overwrite but take special care of `#f`? Like `#f=foo&o=bar&m=baz` ----setObservationId('buf')----> `#f=foo&o=buf`?